### PR TITLE
Improved Cloudbase-Init installation logic

### PR DESCRIPTION
**Pull request by dtoncu**
Changed the logic, so that the method will retry to install Cloudbase-Init until times out or successful installation.
After installation, it is checked for validity. If it's an invalid installation, before retrying, a clean-up is done.
This is a better approach since the installation was sometimes previously unsuccessful and it caused a scenario to fail.

### CODE REVIEW

**Comment by dtoncu** [comment]
2016-09-20 03:17:03,788 - argus - INFO - Trying to install Cloudbase-Init ...
2016-09-20 03:17:03,789 - argus - INFO - Running the downloaded installation script using the installer 'CloudbaseInitSetup_Beta_x64.msi'.
2016-09-20 03:17:03,789 - argus - DEBUG - Executing resource script windows/installCBinit.ps1 with this parameters -installer CloudbaseInitSetup_Beta_x64.msi
2016-09-20 03:17:03,789 - argus - DEBUG - Downloading from https://raw.githubusercontent.com/cloudbase/cloudbase-init-ci/master/argus/resources/windows/installCBinit.ps1 to C:\installCBinit.ps1
2016-09-20 03:17:03,789 - argus - INFO - Running command Invoke-WebRequest -Uri https://raw.githubusercontent.com/cloudbase/cloudbase-init-ci/master/argus/resources/windows/installCBinit.ps1 -OutFile "C:\installCBinit.ps1"...
2016-09-20 03:17:04,705 - argus - INFO - Running command "C:\installCBinit.ps1" -installer CloudbaseInitSetup_Beta_x64.msi...
2016-09-20 03:18:24,557 - argus - INFO - Running command (Get-CimInstance Win32_OperatingSystem).OSArchitecture...
2016-09-20 03:18:25,132 - argus - INFO - Running command powershell "$ENV:ProgramFiles"...
2016-09-20 03:18:25,519 - argus - INFO - Running command powershell "${ENV:ProgramFiles(x86)}"...
2016-09-20 03:18:25,876 - argus - INFO - Running command Test-Path "C:\Program` Files\Cloudbase` Solutions"...
2016-09-20 03:18:26,264 - argus - INFO - Running command dir "C:\Program Files\Cloudbase Solutions\Cloudbase-Init" /b...
2016-09-20 03:18:26,367 - argus - INFO - Checking Cloudbase-Init installation ...
2016-09-20 03:18:26,878 - argus - ERROR - Cloudbase-Init installation failed: Executing command 'cd "C:\\Program Files\\Cloudbase Solutions\\Cloudbase-Init\\Python" ; .\\python.exe -c "import cloudbaseinit"' with encoded Command'powershell -Noninteractive -NoLogo -EncodedCommand YwBkACAAIgBDADoAXABQAHIAbwBnAHIAYQBtACAARgBpAGwAZQBzAFwAQwBsAG8AdQBkAGIAYQBzAGUAIABTAG8AbAB1AHQAaQBvAG4AcwBcAEMAbABvAHUAZABiAGEAcwBlAC0ASQBuAGkAdABcAFAAeQB0AGgAbwBuACIAIAA7ACAALgBcAHAAeQB0AGgAbwBuAC4AZQB4AGUAIAAtAGMAIAAiAGkAbQBwAG8AcgB0ACAAYwBsAG8AdQBkAGIAYQBzAGUAaQBuAGkAdAAiAA==' failed with exit code 1 and output 'Traceback (most recent call last):\r\n  File "<string>", line 1, in <module>\r\nImportError: No module named \'cloudbaseinit\'\r\n'.
2016-09-20 03:18:26,879 - argus - INFO - Running command (Get-CimInstance Win32_OperatingSystem).OSArchitecture...
2016-09-20 03:18:27,372 - argus - INFO - Running command powershell "$ENV:ProgramFiles"...
2016-09-20 03:18:27,665 - argus - INFO - Running command powershell "${ENV:ProgramFiles(x86)}"...
2016-09-20 03:18:27,999 - argus - INFO - Running command Test-Path "C:\Program` Files\Cloudbase` Solutions"...
2016-09-20 03:18:28,397 - argus - INFO - Cleaning up Cloudbase-Init ...
2016-09-20 03:18:28,397 - argus - INFO - Running command Test-Path -PathType Any -Path "C:\Program Files\Cloudbase Solutions\Cloudbase-Init\.."...
2016-09-20 03:18:28,770 - argus - INFO - Running command Test-Path -PathType Container -Path "C:\Program Files\Cloudbase Solutions\Cloudbase-Init\.."...
2016-09-20 03:18:29,187 - argus - DEBUG - Remove directory  C:\Program Files\Cloudbase Solutions\Cloudbase-Init\..
2016-09-20 03:18:29,187 - argus - INFO - Running command Remove-Item -Force -Recurse -Path 'C:\Program Files\Cloudbase Solutions\Cloudbase-Init\..'...
2016-09-20 03:18:31,322 - argus - INFO - Running the downloaded installation script using the installer 'CloudbaseInitSetup_Beta_x64.msi'.
2016-09-20 03:18:31,322 - argus - DEBUG - Executing resource script windows/installCBinit.ps1 with this parameters -installer CloudbaseInitSetup_Beta_x64.msi
2016-09-20 03:18:31,322 - argus - DEBUG - Downloading from https://raw.githubusercontent.com/cloudbase/cloudbase-init-ci/master/argus/resources/windows/installCBinit.ps1 to C:\installCBinit.ps1
2016-09-20 03:18:31,323 - argus - INFO - Running command Invoke-WebRequest -Uri https://raw.githubusercontent.com/cloudbase/cloudbase-init-ci/master/argus/resources/windows/installCBinit.ps1 -OutFile "C:\installCBinit.ps1"...
2016-09-20 03:18:32,474 - argus - INFO - Running command "C:\installCBinit.ps1" -installer CloudbaseInitSetup_Beta_x64.msi...
2016-09-20 03:19:53,065 - argus - INFO - Running command (Get-CimInstance Win32_OperatingSystem).OSArchitecture...
2016-09-20 03:19:53,690 - argus - INFO - Running command powershell "$ENV:ProgramFiles"...
2016-09-20 03:19:54,020 - argus - INFO - Running command powershell "${ENV:ProgramFiles(x86)}"...
2016-09-20 03:19:54,320 - argus - INFO - Running command Test-Path "C:\Program` Files\Cloudbase` Solutions"...
2016-09-20 03:19:54,704 - argus - INFO - Running command dir "C:\Program Files\Cloudbase Solutions\Cloudbase-Init" /b...
2016-09-20 03:19:54,788 - argus - INFO - Checking Cloudbase-Init installation ...
2016-09-20 03:19:55,340 - argus - INFO - Cloudbase-Init was successfully installed!

File: argus/action_manager/windows.py [hunk 1]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+                          delay=util.RETRY_DELAY, command_type=None):
+        return self._client.run_command_with_retry(
+            cmd, count=count, delay=delay, command_type=command_type
+        )[0]
+
+    def check_cbinit_installation(self, python_path):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation ...")
+
+        check_cmd = r'cd "{}" ; .\python.exe -c "import cloudbaseinit"'.format(
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: You don't need to `cd` in there, you could just use the full path.

File: argus/action_manager/windows.py [hunk 2]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
         self.download_resource("windows/installCBinit.ps1",
                                r"C:\installCBinit.ps1")
 
+    def _execute_function(self, cmd, count=util.RETRY_COUNT,
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: This can be named `_execute` to match other implementations in argus , like : https://github.com/cloudbase/cloudbase-init-ci/blob/master/argus/recipes/base.py#L51

File: argus/action_manager/windows.py [hunk 3]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+                          delay=util.RETRY_DELAY, command_type=None):
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: Default `command_type` should be `util.CMD`.

File: argus/action_manager/windows.py [hunk 4]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self, cbinit_path):
+        """Clean up Cloudbase-Init."""
+        LOG.info("Cleaning up Cloudbase-Init ...")
+
+        self.rmdir("{}\\..".format(cbinit_path))
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: Are you sure this is the only clean-up required ?

Comment (micumatei) [inline argus/action_manager/windows.py]: Are you sure this is the only clean-up required ?

Comment (dtoncu) [reply argus/action_manager/windows.py]: I'm sure that with this clean-up it's working and the main goal is reached! :)

File: argus/action_manager/windows.py [hunk 5]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
         self.download_resource("windows/installCBinit.ps1",
                                r"C:\installCBinit.ps1")
 
+    def _execute(self, cmd, count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
```
Comment (alexcoman) [inline argus/action_manager/windows.py:124]: Missing docstring.

File: argus/action_manager/windows.py [hunk 6]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+                 command_type=util.CMD):
+        return self._client.run_command_with_retry(
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: I think a better approach will be:

``` python
stdout, _, _ =  self._client.run_command_with_retry(cmd, count=count, delay=delay,
                                                    command_type=command_type)
return stdout
```

File: argus/action_manager/windows.py [hunk 7]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+            cmd, count=count, delay=delay, command_type=command_type
+        )[0]
+
+    def check_cbinit_installation(self, python_path):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation ...")
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: One dot is enough.

File: argus/action_manager/windows.py (hunk 4 above)
Comment (alexcoman) [reply argus/action_manager/windows.py]: Change `\\..` with `os.path.dirname(`.

File: argus/action_manager/windows.py [hunk 8]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+        """Clean up Cloudbase-Init."""
+        LOG.info("Cleaning up Cloudbase-Init ...")
+
+        self.rmdir("{}\\..".format(cbinit_path))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Run the installation script for Cloudbase-Init."""
+        LOG.info("Trying to install Cloudbase-Init ...")
```
Comment (alexcoman) [inline argus/action_manager/windows.py:168]: Change back to `debug`.

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py (hunk 4 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: `os.path.dirname()` is not working because given path is a windows-path and the platform on which Python runs, in our case, is Linux.

Comment (alexcoman) [reply argus/action_manager/windows.py]: That you can use `ntpath.dirname`.

File: argus/action_manager/windows.py [hunk 9]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+        """Execute until success and return only the standard output."""
+        stdout, _, _ = self._client.run_command_with_retry(
+            cmd, count=count, delay=delay, command_type=command_type)
+        return stdout
+
+    def check_cbinit_installation(self, python_path):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        check_cmd = r'& "{}\python.exe" -c "import cloudbaseinit"'.format(
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: `python_path` shouldn't contain the `python.exe` ?

File: argus/action_manager/windows.py [hunk 10]
```diff
@@ -115,32 +116,76 @@ def get_installation_script(self):
...
+            except exceptions.ArgusError:
+                # This can happen for multiple reasons,
+                # but one of them is the fact that the installer
+                # can't be installed through WinRM on some OSes
+                # for whatever reason. In this case, we're falling back
+                # to use a scheduled task.
+                LOG.debug("Could not install Cloudbase-Init!")
+
+                LOG.info("Deploying Cloudbase-Init using a scheduled task ...")
+                self._deploy_using_scheduled_task(installer)
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: What happens if the `_deploy_using_scheduled_task` will fail ?

**Review by alexcoman — CHANGES_REQUESTED** [review]

**Comment by stefan-caraiman** [comment]
Will cloudbase-init show any strange behaviour if it reinstalls after a cleanup, because I imagine there will be keys left in the registry.
Also, I'm curious, how did you test the patch itself? I'm referring to how you managed to make the installer fail(restarted the instance while installing, deleted files, dropped the connection?)

**Comment by dtoncu** [comment]
@stefan-caraiman , actually, there is no strange behavior.. It works like it was installed for the first time. Regarding how I tested this patch, while Cloudbase-Init installation was in progress, I deleted some its files.

File: argus/action_manager/windows.py (hunk 5 above)
Comment (dtoncu) [reply argus/action_manager/windows.py:124]: Done.

File: argus/action_manager/windows.py (hunk 9 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: Do you mean, this parameter should be renamed as `python_dir` ?

File: argus/action_manager/windows.py (hunk 10 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: In previous implementation, this case was left untreated like a Plan B.
Should I include this method in the new `install_cbinit` flow?

Comment (alexcoman) [reply argus/action_manager/windows.py]: Yes.

File: argus/action_manager/windows.py (hunk 9 above)
Comment (alexcoman) [reply argus/action_manager/windows.py]: That may be a solution or you can pass the path of the python binary.

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py (hunk 5 above)
Comment (alexcoman) [inline argus/action_manager/windows.py]: It's this method really required ?

File: argus/action_manager/windows.py [hunk 11]
```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
...
+            cmd, count=count, delay=delay, command_type=command_type)
+        return stdout
+
+    def check_cbinit_installation(self, python_dir):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        check_cmd = r'& "{}\python.exe" -c "import cloudbaseinit"'.format(
+            python_dir)
+
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: This empty line is not required.

File: argus/action_manager/windows.py [hunk 12]
```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
...
                                r"C:\installCBinit.ps1")
 
+    def _execute(self, cmd, count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
+                 command_type=util.CMD):
+        """Execute until success and return only the standard output."""
+        stdout, _, _ = self._client.run_command_with_retry(
+            cmd, count=count, delay=delay, command_type=command_type)
+        return stdout
+
+    def check_cbinit_installation(self, python_dir):
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: Why `python_dir` it's an argument ?

File: argus/action_manager/windows.py [hunk 13]
```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
...
-                parameters=parameters)
-        except exceptions.ArgusError:
-            # This can happen for multiple reasons,
-            # but one of them is the fact that the installer
-            # can't be installed through WinRM on some OSes
-            # for whatever reason. In this case, we're falling back
-            # to use a scheduled task.
-            LOG.debug("Cannot install, deploying using a scheduled task.")
-            self._deploy_using_scheduled_task(installer)
+        for _ in range(util.RETRY_COUNT):
```
Comment (alexcoman) [inline argus/action_manager/windows.py:175]: This code block should be rewritten because it's hard to follow.

You can have a private method which will receive an action (the method of installing cloudbase-init), it will run that action and after that will check if everything it's okay (and if it's not it will clean up the environment).
If that method will return a bool it will be very easy to follow the workflow of the current method.

@stefan-caraiman @micumatei @ader1990 any other ideas ?

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py [hunk 14]
```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
...
+                try:
+                    self._deploy_using_scheduled_task(installer)
+                except exceptions.ArgusError:
+                    LOG.debug("Could not install Cloudbase-Init "
+                              "using a scheduled task!")
+
+                    LOG.debug("Retrying ...")
+                    continue
+
+            python_dir = introspection.get_python_dir(self._execute)
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: What of there is no `python_dir` ? If we fail installation we might not have Cloudbase-Init directory, and if we don't have one we won't have any python directory.This function `get_python_dir` uses `get_cbinit_dir` that raises an Exception if no Cloudbase-Init directory could be found.

https://github.com/cloudbase/cloudbase-init-ci/blob/master/argus/introspection/cloud/windows.py#L155

**Review by mateimicu — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py [hunk 15]
```diff
@@ -115,34 +117,95 @@ def get_installation_script(self):
...
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Install Cloudbase-Init on the underlying instance."""
+        LOG.info("Trying to install Cloudbase-Init ...")
+
+        for _ in range(util.RETRY_COUNT):
+            for install_method in (self._run_installation_script,
+                                   self._deploy_using_scheduled_task):
+                try:
+                    install_method()
```
Comment (stefan-caraiman) [inline argus/action_manager/windows.py]: Who is this `install_method` ?

File: argus/action_manager/windows.py [hunk 16]
```diff
@@ -115,34 +117,95 @@ def get_installation_script(self):
...
-            # to use a scheduled task.
-            LOG.debug("Cannot install, deploying using a scheduled task.")
-            self._deploy_using_scheduled_task(installer)
-
-    def _deploy_using_scheduled_task(self, installer):
+        self.execute_powershell_resource_script(
+            resource_location='windows/installCBinit.ps1',
+            parameters=parameters)
+
+    def _deploy_using_scheduled_task(self):
```
Comment (stefan-caraiman) [inline argus/action_manager/windows.py]: I think it was alright to simply give the already created `installer` string from the `_run_installation_script` since you already had it formatted there.

Comment (dtoncu) [reply argus/action_manager/windows.py]: You're right.

File: argus/action_manager/windows.py (hunk 15 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: At the first iteration, `install_method` is `_run_installation_script`. At the second iteration, if `_run_installation_script` throws an exception, it will be `_deploy_using_scheduled_task`.

File: argus/action_manager/windows.py [hunk 17]
```diff
@@ -120,34 +121,85 @@ def get_installation_script(self):
...
+            for install_method in (self._run_installation_script,
+                                   self._deploy_using_scheduled_task):
+                try:
+                    install_method(installer)
+                except exceptions.ArgusError as exc:
+                    LOG.debug("Could not install Cloudbase-Init: %s", exc)
+                    self.cbinit_cleanup()
+                else:
+                    if self.check_cbinit_installation():
+                        return
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: It might be just me but this : 

```
if self.check_cbinit_installation():
    return
else:
    break  
```

is hard to follow and looks ugly.

File: argus/action_manager/windows.py [hunk 18]
```diff
@@ -120,34 +121,85 @@ def get_installation_script(self):
...
+    def check_cbinit_installation(self):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        try:
+            python_dir = introspection.get_python_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: This should not be here.

File: argus/action_manager/windows.py [hunk 19]
```diff
@@ -120,34 +121,85 @@ def get_installation_script(self):
...
 
+    def check_cbinit_installation(self):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        try:
+            python_dir = introspection.get_python_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
```
Comment (mateimicu) [inline argus/action_manager/windows.py]: You can fit `delay=util.RETRY_DELAY,` on the first line I think.

**Review by mateimicu — CHANGES_REQUESTED** [review]
This needs good testing.

**Review by mateimicu — CHANGES_REQUESTED** [review]
You have extra commits.Please rebase this PR.

File: argus/action_manager/windows.py (hunk 19 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: Yes, but this does nothing change.

File: argus/action_manager/windows.py (hunk 18 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: Why not? I think, this coding-style approach changes the focus on lambda expression and makes code simplest to follow.

File: argus/action_manager/windows.py (hunk 17 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: Any suggestions ?

File: argus/action_manager/windows.py [hunk 20]
```diff
@@ -120,32 +121,85 @@ def get_installation_script(self):
         self.download_resource("windows/installCBinit.ps1",
                                r"C:\installCBinit.ps1")
 
+    def check_cbinit_installation(self):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        try:
+            python_dir = introspection.get_python_dir(
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: I think that the following code block should be refactored because it's pretty hard to follow what's happen there. Also I think a note comment will be helpful.

File: argus/action_manager/windows.py [hunk 21]
```diff
@@ -120,32 +121,85 @@ def get_installation_script(self):
...
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self):
+        """Clean-up Cloudbase-Init if it's installed."""
+        LOG.info("Cleaning up Cloudbase-Init.")
+
+        try:
+            cbinit_dir = introspection.get_cbinit_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
```
Comment (alexcoman) [inline argus/action_manager/windows.py]: This code block is redundant.

File: argus/action_manager/windows.py (hunk 17 above)
Comment (alexcoman) [reply argus/action_manager/windows.py]: What about the following approach ?

``` python
for _ in range(util.RETRY_COUNT):
    for install_method in (self._run_installation_script,
                           self._deploy_using_scheduled_task):
        try:
            install_method(installer)
        except exceptions.ArgusError as exc:
            LOG.debug("Could not install Cloudbase-Init: %s", exc)            
        else:
            if self.check_cbinit_installation():
                return True
        self.cbinit_cleanup()
return False
```

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py (hunk 21 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: Should I rewrite and use that `_execute()` method instead of lambda expressions?

File: argus/action_manager/windows.py (hunk 17 above)
Comment (dtoncu) [reply argus/action_manager/windows.py]: It's a good approach but it changes the method logic.
This method should try to install **Cloudbase-Init** using `_deploy_using_scheduled_task` method only in case of `_run_installation_script` method raises an exception. If `_run_installation_script` method runs normally but installation was invalid, main method should retry to install **Cloudbase-Init** using the same, `_run_installation_script`, method.

File: argus/resources/windows/schedule_installer.bat [hunk 22]
```diff
@@ -1,4 +1,4 @@
-schtasks /CREATE /TN "cloudbaseinit-installer" /SC ONCE /SD 01/01/2020 /ST 00:00:00 /RL HIGHEST /RU CiAdmin /RP Passw0rd /TR "powershell C:\\installcbinit.ps1  -serviceType %1 -installer %2" /F
+schtasks /CREATE /TN "cloudbaseinit-installer" /SC ONCE /SD 01/01/2020 /ST 00:00:00 /RL HIGHEST /RU CiAdmin /RP Passw0rd /TR "powershell C:\\installcbinit.ps1 -installer %2" /F
```
Comment (stefan-caraiman) [inline argus/resources/windows/schedule_installer.bat:1]: Are you sure this is correct? `C:\\installcbinit.ps1 -installer %2" /F` `%2` won't it refer to a second argument? I think you should replace it to `%1`.

File: argus/tests/action_manager/test_windows.py [hunk 23]
```diff
@@ -0,0 +1,187 @@
+# Copyright 2014 Cloudbase Solutions Srl
```
Comment (stefan-caraiman) [inline argus/tests/action_manager/test_windows.py]: Move this to `argus/unit_tests/action_manager`.

File: argus/resources/windows/schedule_installer.bat (hunk 22 above)
Comment (dtoncu) [reply argus/resources/windows/schedule_installer.bat:1]: This script is called as follows: `schedule_installer.bat -installer CloudbaseInitSetup_*_*.msi`. So, `%1` refers to `"-installer"` and `%2` - `"CloudbaseInitSetup_*_*.msi"`.

File: argus/unit_tests/action_manager/test_windows.py [hunk 24]
```diff
@@ -0,0 +1,216 @@
+# Copyright 2014 Cloudbase Solutions Srl
```
Comment (alexcoman) [inline argus/unit_tests/action_manager/test_windows.py]: Change `2014` with `2016`.

File: argus/unit_tests/action_manager/test_windows.py [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+    import unittest.mock as mock
+except ImportError:
+    import mock
+
+from argus.action_manager import windows as action_manager
+from argus import exceptions
+from argus.introspection.cloud import windows as introspection
+from argus import util
+
+LOG = util.get_logger()
```
Comment (alexcoman) [inline argus/unit_tests/action_manager/test_windows.py]: The logger is not required.

File: argus/unit_tests/action_manager/test_windows.py [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+
+
+class WindowsActionManagerTest(unittest.TestCase):
+    """Tests for windows action manager class."""
+
+    def setUp(self):
+        self._client = mock.MagicMock()
+        self._config = mock.MagicMock()
```
Comment (alexcoman) [inline argus/unit_tests/action_manager/test_windows.py:34]: You can use the config object and mock only the values that should be changed.

File: argus/unit_tests/action_manager/test_windows.py [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+        self._os_type = mock.MagicMock()
```
Comment (alexcoman) [inline argus/unit_tests/action_manager/test_windows.py]: The `_os_type` can be a constant.

File: argus/unit_tests/action_manager/test_windows.py [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+
+        self._action_manager = action_manager.WindowsActionManager(
+            client=self._client, config=self._config, os_type=self._os_type)
+
+        self._util = mock.MagicMock()
```
Comment (alexcoman) [inline argus/unit_tests/action_manager/test_windows.py]: Why it is required ?

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/unit_tests/action_manager/test_windows.py [hunk 29]
```diff
@@ -0,0 +1,275 @@
...
+        cmd = "fake-command"
+
+        if exc:
+            self._client.run_command_with_retry = mock.Mock(side_effect=exc)
+            with self.assertRaises(exc):
+                self._action_manager._execute(
+                    cmd, count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
+                    command_type=util.CMD)
+        else:
+            self._client.run_command_with_retry = \
```
Comment (mateimicu) [inline argus/unit_tests/action_manager/test_windows.py]: This should work : 

```
self._client.run_command_with_retry = mock.Mock(
    return_value=("fake-stdout", "fake-stderr", 0))
```

**Review by mateimicu — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py [hunk 30]
```diff
@@ -120,32 +121,84 @@ def get_installation_script(self):
...
+            self._client.run_remote_cmd(
+                cmd=check_cmd, command_type=util.POWERSHELL)
+        except exceptions.ArgusError as exc:
+            LOG.debug("Cloudbase-Init installation failed: %s", exc)
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self):
```
Comment (alexcoman) [inline argus/action_manager/windows.py:153]: Shouldn't add cleanup for `_deploy_using_scheduled_task` too ?

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/action_manager/windows.py (hunk 30 above)
Comment (dtoncu) [reply argus/action_manager/windows.py:153]: Do you mean, `_deploy_using_scheduled_task` method could need some additional specific clean up?

Comment (mateimicu) [reply argus/action_manager/windows.py:153]: Yes.

**Comment by stefan-caraiman** [comment]
Please fix conflicts.

**Comment by alexcoman** [comment]
One month and over 20 reviews for this PR are way to much for this task.
The PR will be closed.

The task will be assigned back to @stefan-caraiman.
